csharp/src/Google.Protobuf/Collections/RepeatedField.cs

```diff
@@ -645,7 +645,86 @@ void IList.Remove(object value)
                 Remove(t);
             }
         }
-        #endregion        
+        #endregion
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <returns></returns>
+        public Enumerable AsStructEnumerable() => new Enumerable(this);
+
+        /// <summary>
+        /// 
+        /// </summary>
+        public readonly struct Enumerable : IEnumerable<T>
+        {
+            private readonly RepeatedField<T> field;
+
+            internal Enumerable(RepeatedField<T> field)
+            {
+                this.field = field;
+            }
+
+            /// <summary>
+            /// 
+            /// </summary>
+            /// <returns></returns>
+            public Enumerator GetEnumerator() => new Enumerator(field);
+
+            IEnumerator<T> IEnumerable<T>.GetEnumerator() => GetEnumerator();
+
+            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        public struct Enumerator : IEnumerator<T>
+        {
+            private readonly RepeatedField<T> field;
+            private int index;
+
+            internal Enumerator(RepeatedField<T> field)
+            {
+                this.field = field;
+                this.index = -1;
+            }
+
+            /// <summary>
+            /// 
+            /// </summary>
+            public T Current => index == -1 || index >= field.Count
+                ? throw new InvalidOperationException()
+                : field[index];
```

Comment:
Is this check really necessary? The indexer does the exact same check, but throws a slightly different exception. Skipping the check here would create an implicit dependency on that implementation, but that seems fine given that it belongs to the same class.

Reply:
I think it's worth making sure we throw the right exception, personally.